### PR TITLE
YOLOv5 transfer learning and hyperparameter optimisation using Amazon SageMaker

### PR DESCRIPTION
**Pull request by OmarKhayyam**
*Issue #, if available:* None.

*Description of changes:* Added Jupyter notebook, and other artifacts (helper code, and sample output) that enable the use of Amazon SageMaker for training YOLOv5 on a custom dataset, and also enables hyperparameter optimization.

*Testing done:* Ran the notebook on a SageMaker Notebook instance and Studio environment

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-sagemaker-examples/blob/master/CONTRIBUTING.md) doc and adhered to the example notebook best practices
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-sagemaker-examples/blob/master/README.md)
- [x] I have tested my notebook(s) and ensured it runs end-to-end
- [x] I have linted my notebook(s) and code using `black-nb -l 100 {path}/{notebook-name}.ipynb`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: advanced_functionality/pytorch_yolov5_training_and_hpo/yolov5/training-wrapper.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+        initialstr += " --freeze " + str(arguments.freeze)
+    if arguments.epochs:
+        initialstr += " --epochs " + str(arguments.epochs)
+    if arguments.patience:
+       initialstr += " --patience " + str(arguments.patience) 
+    if arguments.name:
+       initialstr += " --name " + str(arguments.name) 
+    ## Our default hyperparameter values are sourced from yolov5/data/hyps/hyp.scratch-low.yaml
+    ## We then override what we want changed.
+    defhyps = open('/yolov5/data/hyps/hyp.scratch-low.yaml')
```
[the remark this answers is not in the document]
Comment (OmarKhayyam) [reply advanced_functionality/pytorch_yolov5_training_and_hpo/yolov5/training-wrapper.py]: Made the suggested changes to the training-wrapper.py.

File: advanced_functionality/pytorch_yolov5_training_and_hpo/yolov5/training-wrapper.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+def createHypsFile(opts):
+    ## Create a new file that we can use
+    if os.path.exists(SMHYPFILE):
+        os.remove(SMHYPFILE)
+    hypfile = open(SMHYPFILE,'w') # New file
+    args = yaml.dump(opts,hypfile)    
+    hypfile.close()
+    return " --hyp " + SMHYPFILE
+
+def buildCMD(arguments):
```
[the remark this answers is not in the document]
Comment (OmarKhayyam) [reply advanced_functionality/pytorch_yolov5_training_and_hpo/yolov5/training-wrapper.py:41]: Made the suggested changes to the training-wrapper.py.

File: advanced_functionality/pytorch_yolov5_training_and_hpo/yolov5/training-wrapper.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+            else:
+                raise
+    print("**Data Configuration**:")
+    with open(DATACFGFILE) as f:
+        print(f.read())
+    print("**Hyperparameters Configuration**:")
+    with open(SMHYPFILE) as f:
+        print(f.read())
+    print("**Executing** : {}".format(cmd))
+    status = os.system(cmd) ## We will replace this with subprocess.Popen once we see that this works
```
[the remark this answers is not in the document]
Comment (OmarKhayyam) [reply advanced_functionality/pytorch_yolov5_training_and_hpo/yolov5/training-wrapper.py]: Made the suggested changes to the training-wrapper.py.

File: advanced_functionality/pytorch_yolov5_training_and_hpo/yolov5/training-wrapper.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    initialstr = "/opt/conda/bin/python3.8 /yolov5/train.py --project /opt/ml/model --cache "
+
+    argls = (arguments.__dict__).items()    
+    
+    for arg in argls:
+        if arg[1] is None or arg[0] in ['train','val']:
+            continue
+        elif arg[1] is not None and arg[0] in non_hyps:
+            if arg[0] == 'batchsize':
+                initialstr += " --batch " + str(arg[1])
```
[the remark this answers is not in the document]
Comment (OmarKhayyam) [reply advanced_functionality/pytorch_yolov5_training_and_hpo/yolov5/training-wrapper.py]: changed code to fix this.

File: advanced_functionality/pytorch_yolov5_training_and_hpo/transfer_learning_and_hpo_yolov5.ipynb [hunk 5]
```diff
@@ -0,0 +1,1379 @@
+{
```
Comment (atqy) [inline advanced_functionality/pytorch_yolov5_training_and_hpo/transfer_learning_and_hpo_yolov5.ipynb:1]: <p>Stick with present tense -&gt; "This notebooks takes you...", "You use AWS Deep Learning Containers...", "You also use Amazon SageMaker's Automated Model Tuning"</p><p> </p><p>This applies to everything in this notebook as well</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3601/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (atqy) [inline advanced_functionality/pytorch_yolov5_training_and_hpo/transfer_learning_and_hpo_yolov5.ipynb:1]: <p>Did you submit a request for dataset review by legal?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3601/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (atqy) [inline advanced_functionality/pytorch_yolov5_training_and_hpo/transfer_learning_and_hpo_yolov5.ipynb:1]: <p>Break up sentence -&gt; "file that has to follow a specific structure. In this structure "</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3601/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='75'/>

Comment (OmarKhayyam) [reply advanced_functionality/pytorch_yolov5_training_and_hpo/transfer_learning_and_hpo_yolov5.ipynb:1]: No, i haven't. How do I go about it? Any guidance on a Wiki?

Comment (OmarKhayyam) [reply advanced_functionality/pytorch_yolov5_training_and_hpo/transfer_learning_and_hpo_yolov5.ipynb:1]: <p>Changed all to present tense. Also stayed with second person singular. Do let me know if this is OK.</p>

Comment (OmarKhayyam) [reply advanced_functionality/pytorch_yolov5_training_and_hpo/transfer_learning_and_hpo_yolov5.ipynb:1]: <p>Just for the record, the request to legal was submitted, and legal approved it. TT URL: <a href="https://t.corp.amazon.com/D57175372/overview" target="_blank">https://t.corp.amazon.com/D57175372/overview</a></p><p> </p>

Comment (OmarKhayyam) [reply advanced_functionality/pytorch_yolov5_training_and_hpo/transfer_learning_and_hpo_yolov5.ipynb:1]: <p>Broke up the sentence, this is what it looks like now, let me know if this is OK:</p><p> </p><p>Whether you run a training job, or perform automated model tuning (AMT), you need to create a <code>model.tar.gz</code> file that has to follow a specific structure. In this structure you will need to have the model, in this case the <code>best.pt</code> file. This is available in the <code>weights</code> directory in your <code>model.tar.gz</code> file. For a deeper understanding of how to create the model directory structure refer to <a href="https://sagemaker.readthedocs.io/en/stable/frameworks/pytorch/using_pytorch.html#deploy-pytorch-models" target="_blank">Use PyTorch with SageMaker SDK</a>.</p>

**Review by atqy — APPROVED** [review]

File: advanced_functionality/pytorch_yolov5_training_and_hpo/transfer_learning_and_hpo_yolov5.ipynb (hunk 5 above)
Comment (atqy) [reply advanced_functionality/pytorch_yolov5_training_and_hpo/transfer_learning_and_hpo_yolov5.ipynb:1]: <p>Yeah, it looks good</p>

File: advanced_functionality/pytorch_yolov5_training_and_hpo/transfer_learning_and_hpo_yolov5.ipynb [hunk 6]
```diff
@@ -0,0 +1,1633 @@
...
```
[the remark this answers is not in the document]
Comment (OmarKhayyam) [reply advanced_functionality/pytorch_yolov5_training_and_hpo/transfer_learning_and_hpo_yolov5.ipynb:1]: Changed the first cell to the following, also fixed a spelling mistake in the next cell. Reduced verbosity and broke up some long sentences. Let me know if you need me to fix anything else. This is what the cell looks like now:

In Computer Vision (CV), object detection is a technique where a model predicts the presence of objects and locates them in an image using bounding boxes. YOLO (You Only Look Once) belongs to the family of models used for object detection.

There are two approaches that are commonly adopted for this task. A two-step approach and a single shot approach. [YOLO](https://arxiv.org/abs/1506.02640) is a single shot approach to object detection.

**Review by atqy — APPROVED** [review]

**Pull request merged**